### PR TITLE
Implement JPEG XL support

### PR DESCRIPTION
**Pull request by frozolotl**
This PR implements JPEG XL support.
JPEG XL images can be imported and they do display correctly.

However, I wrote this code mostly as a PoC.
That means, the code isn't _great_ and CMYK images can't be imported yet.
Also, tests are missing.
I also didn't check if `jxl-oxide` was the best library to use and how complete it actually is. EDIT: seems pretty good (https://github.com/web-platform-tests/interop/issues/430#issuecomment-1811972676)

If we decide that this feature is one we want, I'll gladly finish this work.
For now though, I'll keep it a draft PR.

### CODE REVIEW

**Comment by Enivex** [comment]
Some of the comments in this issue describe why JPEG XL is the future better than I am able to

https://github.com/web-platform-tests/interop/issues/430

TL;DR: It basically beats every existing image format at everything.

Support is currently limited, but with Firefox nightly having it, Safari recently adding support, [Adobe adopting it](https://github.com/web-platform-tests/interop/issues/430#issuecomment-1753109905) and [Google working on it](https://github.com/web-platform-tests/interop/issues/430#issuecomment-1776762287), it'll likely become "the" format.

I think it would make sense for typst to adopt it both for image inclusion, and as a render target.

**Comment by Dherse** [comment]
If we can wait for deferred image encoding before that'd be nice, looks like a purge to merge the two 😂

**Comment by Enivex** [comment]
FYI there is actually a pure rust encoder for JPEG XL

https://crates.io/crates/zune-jpegxl

Only lossless, but that is the most interesting for typst's use case anyway.

**Comment by laurmaedje** [comment]
I just checked and aside from the bunch of crates it adds, this also adds ~700KB binary size overhead for optimized WASM. Not sure what to do with that information, but wanted to leave it here.

The WASM blob grew from 11.5MB in 0.6 to 17MB in 0.9 ... We serve it as compressed brotli since a while which brings it down to ~5MB actual download.

**Comment by laurmaedje** [comment]
So, since PDF doesn't support JPEG XL, the only benefit to supporting it would be that users don't have to convert their JPEG XL files to normal JPEG before passing them to Typst. I think it makes sense to have common formats built-in, but is it common? Do people actually have JPEG XL files? I've never seen one in practice.

**Comment by roland-5** [comment]
I converted my graphics and photos (jpg, png) to jpeg-xl. They take up much less space and if it was a jpeg file, thanks to bit-to-bit compatibility, I can restore them as jpeg files with the same checksum. However, you must remember that jxl is not yet in version 1.0 and even though it has already proven to be a better format than the others, there are a few things missing or problematic (the time needed to load highly compressed photos is terrible...). This is certainly the future graphics format of the web, but it's a pity that Google, wanting to save avif and webp, blocked further jxl expansion, and Mozilla doesn't want to get ahead of the curve and the support is hidden behind a flag in about:config. The lack of jxl support for PDFs is a big drawback and I have no idea what the development of PDF as a format looks like.

**Comment by Enivex** [comment]
> So, since PDF doesn't support JPEG XL, the only benefit to supporting it would be that users don't have to convert their JPEG XL files to normal JPEG before passing them to Typst. I think it makes sense to have common formats built-in, but is it common? Do people actually have JPEG XL files? I've never seen one in practice.

It's not common yet, because it's very new.

Note that JPEG XL doesn't just replace JPEG. It also replaces PNG.

**Comment by laurmaedje** [comment]
Since the format isn't yet in widespread use and this would introduce a considerable amount of new crates into the dependency tree and a good chunk of binary size, too, I would rather not merge this for now. I don't think there is enough benefit so far. As discussed on Discord, this functionality could be provided by a plugin for the time being. Thanks for the work still!

**Comment by squeaktoy** [comment]
I'd be interested in such a plugin

**Comment by extua** [comment]
Update tracking [a similar comment](https://github.com/typst/typst/pull/2925#issuecomment-1861955405) I made in the WebP support PR (#2925), there was news from Peter Wyatt of the PDF Association this month:

> The members of the [Imaging Model Technical Working Group](https://pdfa.org/community/imaging-model-twg/) considered various codecs/formats and have agreed on JPEG-XL as the primary way forward at this time.

https://github.com/pdf-association/pdf-issues/discussions/248#discussioncomment-11431638

So, of all the next-gen image formats, JPEG-XL is now the one being considered for adoption in a future PDF standard.

**Comment by DDjivan** [comment]
Is there any plugin or update to having JPEG XL support?

**Comment by frabera** [comment]
I created a WASM plugin to load JPEG XL files: https://github.com/frabera/jxltypst

It is a proof of concept, it decodes the image using jxl-rs https://github.com/libjxl/jxl-rs, the new rust native decoder implementation, then pass the raw `rgb8` or `rgba8` data to the native image function.

I think it re-encodes it in PNG. I also evaluated to re-encode the data in JPEG as I think that Typst doesn't re-encode it and just embeds it to the PDF, but it would be lossy and slower. Do you have any opinion about this?

**Comment by DDjivan** [comment]
> I created a WASM plugin to load JPEG XL files: https://github.com/frabera/jxltypst
> 
> It is a proof of concept, it decodes the image using jxl-rs https://github.com/libjxl/jxl-rs, the new rust native decoder implementation, then pass the raw `rgb8` or `rgba8` data to the native image function.
> 
> I think it re-encodes it in PNG. I also evaluated to re-encode the data in JPEG as I think that Typst doesn't re-encode it and just embeds it to the PDF, but it would be lossy and slower. Do you have any opinion about this?

That's pretty cool! I wanted to try it following the example in your git repository's README.md, but… 

```typst
#import "@preview/jxltypst:0.6.0": image-jxl
```

Did you just copy and paste from https://github.com/nineff/grayness , changed the name of the package… without actually publishing it on Typst Universe? 

Edit: Also, only a single commit for everything? How? D:

**Comment by frabera** [comment]
Yeah, the version is wrong sorry! Anyway the package is not in universe, so you need to download the package (it is already built, it's the typst folder) and install it locally or import `lib.typ`

**Comment by frabera** [comment]
The JPEG XL loader plugin is now available in universe: https://typst.app/universe/package/jxl-loader
